### ISSUE TITLE
Question: Does profiling API support Linux?

### ISSUE THREAD

**Issue by noahfalk** [dotnet/coreclr#18478]
This was originally posted by @dotnetjt . To avoid pinging the subscribers to our announcement issue (#15136) with a potentially ongoing discussion I moved it here:

> Hi @noahfalk - just wondering if the profiling API support for Linux went out in 2.1. I've been coding off and on, still working against a 2.1 preview.
> I fired up a new Ubuntu server and installed the 2.1 runtime, and set up environment to point to my (portable) profiler library and it doesn't even attempt to load. I'm pretty sure it's not me as ldd shows that I'm not missing any libraries, so all of my static linking is fine.
> Only other thing I can think of is that 2.1 doesn't support it yet.

Glad you reached out and sorry its causing you trouble! Linux is fully supported ([profiler status page](https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/profiling-api-status.md)) so this should work.

If you have a repro I could help diagnose the problem, or if you are set up to [debug the runtime](https://github.com/dotnet/coreclr/blob/master/Documentation/building/debugging-instructions.md) (you don't need SOS, regular C++ debugging is fine) you could set a breakpoint at https://github.com/dotnet/coreclr/blob/master/src/vm/profilinghelper.cpp#L690 and then see what fails.

**Comment by dotnetjt** [comment]
Thanks for the quick reply Noah. Don't want to take too much of your time, so just a few quick questions.

Naturally, I spent some time staring at my env vars, wondering if I just had a typo, as it seems like the runtime just completely skips over my profiler. Wouldn't be the first time.

I would know exactly how to debug this on the windows side, so this is new to me.
1. To debug the runtime, does my .net app I'm testing against need to be using my locally built runtime? Or is using the installed 2.1 rtw version fine? (Subquestion, if the answer is I just need the source, master branch still ok?)

2. Is the runtime logging the failures in loading a profiler anywhere, similar to how it logs to event viewer in windows?

**Comment by davmason** [comment]
@dotnetjt on linux we pipe all the profiler messages to OutputDebugString, which in the pal is turned off by default. I believe you can enable it by setting the environment variable `PAL_OUTPUTDEBUGSTRING=1`, but it's been a while since I used it.

It's possible to debug the release version of the runtime, but it's much easier to do it with a locally built debug version, then sources and symbols will automatically work.

If you are set up to debug the runtime, I've found it faster in the past to just set a breakpoint on `ProfilingAPIUtility::LoadProfiler` and see why it's failing. 

https://github.com/dotnet/coreclr/blob/66ff5d7ed68c050ba53968c2ecf88a6c9a2a3224/src/vm/profilinghelper.cpp#L869-L1274

**Comment by noahfalk** [comment]
Yeah it is referring to assemblies from the CoreFX build. We should improve the docs. You've actually got a number of ways you can do this:

1) Write a test app that has no dependencies other than System.Private.CoreLib.dll and launch it with corerun.exe right in the coreclr repo build directory. Thankfully you don't need much to test attaching a profiler:

```
using System;
public static class Program {
     public static int Main(string[] args) {
         return 99;
    }
}

csc /r:System.Private.CoreLib.dll /nostdlib Program.cs
corerun.exe Program.exe
```

2) Install a .NET daily build that is sufficiently close to the commit in the CoreCLR repo and then copy the coreclr binaries you built into the install directory. This is my little script for windows and Linux is the same sans a few adjustments in filenames:

```
set source=F:\github\coreclr\bin\Product\Windows_NT.x64.Release
set target="C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.1.0"
copy %source%\corerun.exe %target%\corerun.exe
copy %source%\coreclr.dll %target%\coreclr.dll
copy %source%\clrjit.dll %target%\clrjit.dll
copy %source%\mscordaccore.dll %target%\mscordaccore.dll
copy %source%\sos.dll %target%\sos.dll
copy %source%\sos.NETCore.dll %target%\sos.NETCore.dll
copy %source%\clretwrc.dll %target%\clretwrc.dll
copy %source%\system.private.corelib.dll %target%\system.private.corelib.dll
copy %source%\dbgshim.dll %target%\dbgshim.dll
copy %source%\mscordbi.dll %target%\mscordbi.dll
```

3) You can build the CoreFX repo, then copy assemblies that you want to use into the same directory as coreclr.dll. The corerun.exe host serves the same role as dotnet.exe but it has an extremely simple assembly search algorithm so it expects all native dlls and managed assemblies to be co-located in the same folder.

**Comment by dotnetjt** [comment]
That all makes a lot more sense and this is getting adding to my "scrapbook of things I don't remember."

But, you know what I just discovered? Mistake in my path. This is what I get for staring at a console too long late at night. Sorry for the fire drill, but at least I learned some new stuff! 

Btw, using ```PAL_OUTPUTDEBUGSTRING=1``` did work as I thought it would, and gave me messages around profiler loading, so that's awesome too.

Thanks again guys!
